### PR TITLE
Add support for custom headers for improved proxy support.

### PR DESCRIPTION
**Pull request by cmelchior**
Closes https://github.com/realm/realm-java-private/issues/236

This PR improves proxy support for the Realm Object Server by adding the following API's:

```
// SyncManager
SyncManager.addCustomRequestHeader(String header, String headerValue);
SyncManager.addCustomRequestHeader(String header, String headerValue, String host);
SyncManager.addCustomRequestHeaders(Map<String, String> headers);
SyncManager.addCustomRequestHeaders(Map<String, String> headers, String host);
SyncManager.setAuthorizationHeaderName(String headerName);
SyncManager.setAuthorizationHeaderName(String headerName, String host);

// SyncConfiguration.Builder
SyncConfiguration.Builder urlPrefix(String urlPrefix);
```
The API mirrors the one available in Swift + adds the capability to define headers for selected hosts only. 

See https://docs.realm.io/platform/guides/learn-realm-sync-and-integrate-with-a-proxy#adding-a-custom-proxy

**TESTING**:

```
Sync protocol: From Logcat with different auth header + prefix
08-29 10:01:02.059 10666-10692/io.realm.test V/REALM_SYNC: Connection[1]: HTTP request =
    GET /foo/%2Fdefault%2F__partial%2F516d643ff4363a8eecb2e6c16aa84f75%2Fbb54a0f71d1682d8dbc7cf0bca13f52473a48e6a HTTP/1.1
    Connection: Upgrade
    Host: 127.0.0.1
    Sec-WebSocket-Key: hVzyMIEtInK3V0tUc7yRjg==
    Sec-WebSocket-Protocol: io.realm.sync.24
    Sec-WebSocket-Version: 13
    TestAuth: Realm-Access-Token version=1 token="eyJhcHBfaWQiOiJpby5yZWFsbS50ZXN0IiwiaWRlbnRpdHkiOiI1MTZkNjQzZmY0MzYzYThlZWNiMmU2YzE2YWE4NGY3NSIsImFjY2VzcyI6WyJkb3dubG9hZCIsInVwbG9hZCJdLCJzYWx0IjoiNmVkMDBjZTkiLCJleHBpcmVzIjoxNTM1NTM2ODgyLCJwYXRoIjoiL2RlZmF1bHQvX19wYXJ0aWFsLzUxNmQ2NDNmZjQzNjNhOGVlY2IyZTZjMTZhYTg0Zjc1L2JiNTRhMGY3MWQxNjgyZDhkYmM3Y2YwYmNhMTNmNTI0NzNhNDhlNmEiLCJzeW5jX2xhYmVsIjoiZGVmYXVsdCJ9:AGmdmj8PtEIpFjuPNZgZFegzs0D6cOAg/EfgGnwppQgt83GrUDrRjm/wtouBGv94EF+rYavz29B7K+YkN56qGntCumUrwAEHmZURVmebjdk4gxX0W17F9yRM7kLt3OX7928iCaIAjagnWpmvGZXtead7QYqzm+N1HFr9Ayw7mnQf95VKztBm+FknsGxG/Ctercx8nr+GYWYFyhGUZS6kG40eZkoACvyIqc3sxWsjlfmIPpG7NMbQuICovfQhs2L4Sk2clJkwNtnwcUpUsz0peGK4QD4KG+gN7akrTWbTQBCTMl8/ZTZdXNQsFeODv1/ZLWqjMaINw9bol2BmKc/cxg=="
    Upgrade: websocket

```

Login using POST, from Charles (custom headers): 
image

**TODO**
- [x] Add public API's
- [x] Unit tests
- [x] Integration tests (if possible)

### CODE REVIEW

**Comment by cmelchior** [comment]
Ready for review. The tests failing parse locally and are not related to this PR. I'm looking into issues with the CI devices.

File: realm/realm-library/src/main/cpp/io_realm_internal_OsRealmConfig.cpp [hunk 1]
```diff
@@ -333,13 +332,32 @@ JNIEXPORT jstring JNICALL Java_io_realm_internal_OsRealmConfig_nativeCreateAndSe
...
+        if (j_custom_auth_header_name) {
+            JStringAccessor custom_auth_header_name(env, j_custom_auth_header_name);
+            config.sync_config->authorization_header_name = realm::util::Optional<std::string>(custom_auth_header_name);
+        }
+
+        if (j_custom_headers_array) {
+            jsize count = env->GetArrayLength(j_custom_headers_array);
+            for (int i = 0; i < count; i = i + 2) {
+                JStringAccessor key(env, (jstring) env->GetObjectArrayElement(j_custom_headers_array, i));
+                JStringAccessor value(env, (jstring) env->GetObjectArrayElement(j_custom_headers_array, i + 1));
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply realm/realm-library/src/main/cpp/io_realm_internal_OsRealmConfig.cpp]: Yes, but the invariant is guaranteed from the Java side and this block is encapsulated in a try/catch if it by chance breaks.

File: realm/realm-library/src/objectServer/java/io/realm/SyncManager.java [hunk 2]
```diff
@@ -276,6 +283,148 @@ public static synchronized SyncSession getOrCreateSession(SyncConfiguration sync
...
+        authServer.addHeader(headerName, headerValue, null);
+        globalCustomHeaders.put(headerName, headerValue);
+    }
+
+    /**
+     * Adds an extra HTTP header to append to every request to a Realm Object Server.
+     *
+     * @param headerName the name of the header.
+     * @param headerValue the value of header.
+     * @param host if this is provided, the this header will only be used on this particular host.
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/SyncManager.java]: remove `the`

File: realm/realm-library/src/objectServer/java/io/realm/SyncManager.java [hunk 3]
```diff
@@ -276,6 +283,148 @@ public static synchronized SyncSession getOrCreateSession(SyncConfiguration sync
...
+        String host = objectServerUrl.getHost().toLowerCase(Locale.US);
+        String hostRestrictedHeader = hostRestrictedAuthorizationHeaderName.get(host);
+        return (hostRestrictedHeader != null) ? hostRestrictedHeader : globalAuthorizationHeaderName;
+    }
+
+    /**
+     * Returns all the custom headers added to requests to the given url.
+     *
+     * @return all defined custom headers used when making http requests to the given url.
+     * f
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/SyncManager.java]: remove

File: realm/realm-library/src/objectServer/java/io/realm/internal/network/OkHttpAuthenticationServer.java [hunk 4]
```diff
@@ -237,9 +303,28 @@ private UpdateAccountResponse updateAccount(URL updateAccountUrl, String request
                 .addHeader("Content-Type", "application/json")
                 .addHeader("Accept", "application/json");
 
+        // Add custom headers used by all hosts
+        for (Map.Entry<String, String> entry : customHeaders.get("").entrySet()) {
+            builder.addHeader(entry.getKey(), entry.getValue());
+        }
+
+        // add custom headers used by specific host (may override g
```
Comment (nhachicha) [inline realm/realm-library/src/objectServer/java/io/realm/internal/network/OkHttpAuthenticationServer.java]: may override global)

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/SyncedRealmIntegrationTests.java [hunk 5]
```diff
@@ -346,4 +347,96 @@ public void defaultRealm() throws InterruptedException {
...
+                })
+                .build();
+
+        RealmLog.setLevel(LogLevel.ALL);
+        RealmLog.add((level, tag, throwable, message) -> {
+            if (tag.equals("REALM_SYNC")
+                    && message.contains("GET /foo/%2Fdefault%2F__partial%")
+                    && message.contains("TestAuth: Realm-Access-Token version=1")
+                    && message.contains("Test: test")) {
+                looperThread.testComplete();
```
Comment (nhachicha) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/SyncedRealmIntegrationTests.java]: remove the custom log before completing

**Pull request merged**